src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -5,34 +5,650 @@
+            {
+                return (uint)Bmi1.X64.TrailingZeroCount(value);
+            }
+
+            // Instead of writing a 64-bit function,
+            // we use the 32-bit function twice.
+
+            uint hv = (uint)(value >> 32); // High-32
+            uint bv = (uint)value; // Low-32
+
+            uint h, b;
+            if (Bmi1.IsSupported)
+            {
+                h = Bmi1.TrailingZeroCount(hv);
+                b = Bmi1.TrailingZeroCount(bv);
+            }
+            else
+            {
+                long hi = hv & -hv;
+                long bi = bv & -bv;
+
+                hi %= 37;
+                bi %= 37;
+
+                // long.MaxValue % 37 is always in range [0 - 36] so we use Unsafe.AddByteOffset to avoid bounds check
+                ref byte tz = ref MemoryMarshal.GetReference(TrailingZeroCountUInt32);
+
+                h = Unsafe.AddByteOffset(ref tz, (IntPtr)(int)hi);
+                b = Unsafe.AddByteOffset(ref tz, (IntPtr)(int)bi); // Use warm cache
+            }
+
+            // Keep h iff b==32
+            uint mask = b & ~32u; // Zero 5th bit (32)
+            mask = IsZero(mask);  // mask == 0 ? 1 : 0
+            h = mask * h;
+
+            return b + h;
+        }
+
+        #endregion
+
+        #region ExtractBit
+
+        // For bitlength N, it is conventional to treat N as congruent modulo-N
+        // under the shift operation.
+        // So for uint, 1 << 33 == 1 << 1, and likewise 1 << -46 == 1 << +18.
+        // Note -46 % 32 == -14. But -46 & 31 (0011_1111) == +18. So we use & not %.
+        // Software & hardware intrinsics already do this for uint/ulong, but
+        // we need to emulate for byte/ushort.
+
+        /// <summary>
+        /// Reads whether the specified bit in a mask is set.
+        /// Similar in behavior to the x86 instruction BT.
+        /// </summary>
+        /// <param name="value">The mask.</param>
+        /// <param name="bitOffset">The ordinal position of the bit to read.
+        /// Any value outside the range [0..7] is treated as congruent mod 8.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ExtractBit(byte value, int bitOffset)
+        {
+            int shft = bitOffset & 7;
+            uint mask = 1U << shft;
+
+            return (value & mask) != 0;
+        }
+
+        /// <summary>
+        /// Reads whether the specified bit in a mask is set.
+        /// Similar in behavior to the x86 instruction BT.
+        /// </summary>
+        /// <param name="value">The mask.</param>
+        /// <param name="bitOffset">The ordinal position of the bit to read.
+        /// Any value outside the range [0..31] is treated as congruent mod 32.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ExtractBit(uint value, int bitOffset)
+        {
+            // TODO: Not sure if there is a suitable (exposed) intrinsic for this
```

Comment:
Question

Reply:
No. The software fallback is likely good enough.

The closest is probably `bt` (bit test), but there isn't an intrinsic for it and it isn't always a clear win.